// @ng-dynamic-forms/core/esm2015/lib/model/select/dynamic-select.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ɵlooseIdentical as looseIdentical } from "@angular/core";
import { DynamicOptionControlModel } from "../dynamic-option-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isFunction } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_SELECT = "SELECT";
/**
 * @record
 * @template T
 */
export function DynamicSelectModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.compareWithFn;
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.filterable;
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.multiple;
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.placeholder;
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.prefix;
    /** @type {?|undefined} */
    DynamicSelectModelConfig.prototype.suffix;
}
/**
 * @template T
 */
export class DynamicSelectModel extends DynamicOptionControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SELECT;
        this.compareWithFn = isFunction(config.compareWithFn) ? config.compareWithFn : looseIdentical;
        this.filterable = isBoolean(config.filterable) ? config.filterable : false;
        this.multiple = isBoolean(config.multiple) ? config.multiple : false;
        this.placeholder = config.placeholder || "";
        this.prefix = config.prefix || null;
        this.suffix = config.suffix || null;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    select(...indices) {
        /** @type {?} */
        let value = this.multiple ? indices.map((/**
         * @param {?} index
         * @return {?}
         */
        index => this.get(index).value)) : this.get(indices[0]).value;
        this.valueUpdates.next(value);
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicSelectModel.prototype, "filterable", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicSelectModel.prototype, "multiple", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSelectModel.prototype, "placeholder", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSelectModel.prototype, "prefix", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSelectModel.prototype, "suffix", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSelectModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicSelectModel.prototype.compareWithFn;
    /** @type {?} */
    DynamicSelectModel.prototype.filterable;
    /** @type {?} */
    DynamicSelectModel.prototype.multiple;
    /** @type {?} */
    DynamicSelectModel.prototype.placeholder;
    /** @type {?} */
    DynamicSelectModel.prototype.prefix;
    /** @type {?} */
    DynamicSelectModel.prototype.suffix;
    /** @type {?} */
    DynamicSelectModel.prototype.type;
}
