// @ng-dynamic-forms/core/esm5/lib/model/misc/dynamic-form-control-relation.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var MATCH_DISABLED = "DISABLED";
/** @type {?} */
export var MATCH_ENABLED = "ENABLED";
/** @type {?} */
export var MATCH_HIDDEN = "HIDDEN";
/** @type {?} */
export var MATCH_OPTIONAL = "OPTIONAL";
/** @type {?} */
export var MATCH_REQUIRED = "REQUIRED";
/** @type {?} */
export var MATCH_VISIBLE = "VISIBLE";
/** @type {?} */
export var AND_OPERATOR = "AND";
/** @type {?} */
export var OR_OPERATOR = "OR";
/**
 * @record
 */
export function DynamicFormControlCondition() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFormControlCondition.prototype.id;
    /** @type {?|undefined} */
    DynamicFormControlCondition.prototype.rootPath;
    /** @type {?|undefined} */
    DynamicFormControlCondition.prototype.status;
    /** @type {?|undefined} */
    DynamicFormControlCondition.prototype.value;
}
/**
 * @record
 */
export function DynamicFormControlRelation() { }
if (false) {
    /** @type {?} */
    DynamicFormControlRelation.prototype.match;
    /** @type {?|undefined} */
    DynamicFormControlRelation.prototype.operator;
    /** @type {?} */
    DynamicFormControlRelation.prototype.when;
}
